### PR TITLE
Shifting labels for causal LM when using label smoother

### LINKED ISSUE #17960: Suggestion for introducing "shift_labels" argument for Trainer
### Feature request

Add an argument to determine shifting the `labels` or not.

In [TrainingArguments](https://github.com/huggingface/transformers/blob/v4.20.1/src/transformers/training_args.py#L104) class, an argument named `shift_labels` should be added. 

During training, at [here](https://github.com/huggingface/transformers/blob/v4.20.1/src/transformers/models/gpt2/modeling_gpt2.py#L1073) and [here](https://github.com/huggingface/transformers/blob/d0acc9537829e7d067edbb791473bbceb2ecf056/src/transformers/models/gpt2/modeling_gpt2.py#L1280), `model` must check both `labels is not None` and `self.shift_labels is True`

e.g. 
```
if labels is not None and self.shift_labels:          # changed
            # Shift so that tokens < n predict n
            shift_logits = lm_logits[..., :-1, :].contiguous()
            shift_labels = labels[..., 1:].contiguous()
```

Default values for `shift_labels` is `False`, except for causal language models such as `GPT2PreTrainedModel` 

Related to gpt2 : @patil-suraj  and trainer @sgugger

### Motivation

In the current state of the code, the shifting of `labels` for training GPT2LMHeadModel is changing under the use of `label_smoothing`, which I assume is unintended.

Specifically, training a GPT2LMHeadModel with `args.label_smoothing_factor==0` (which is default), the [code](https://github.com/huggingface/transformers/blob/d0acc9537829e7d067edbb791473bbceb2ecf056/src/transformers/models/gpt2/modeling_gpt2.py#L1075) shifts the `labels` and computes the loss inside the `model.forward()`. 
This assumes that `labels` have not been shifted to be properly aligned with corresponding `input_ids`.

However, if I train GPT2LMHeadModel with `args.label_smoothing_factor > 0`, then the loss is computed [here](https://github.com/huggingface/transformers/blob/692e61e91a0b83f5b847902ed619b7c74c0a5dda/src/transformers/trainer.py#L2384), inside the `compute_loss()` function of the `Trainer`. 
This part assumes `labels` are already shifted, and does not proceed to shift the labels.

I believe whether to shift `labels` or not should be explicitly determined by its own argument, not by another argument like `label_smoothing_factor`. In my case, our team was very frustrated that our training results were totally different by only changing the `label_smoothing` with same given `labels` and  `input_ids`. 
The reason was due to the misalignment of  `labels` and  `input_ids` when turning on the `label_smoothing`.

### Your contribution

I'm willing to make PR after your confirmation.

### PR DESCRIPTION
**Pull request by seungeunrho**
# What does this PR do?

Fixes #17960 

When training CausalLM such as GPT2, loss is computed within model's `foward()` function and `labels` are shifted internally. However, if label smoothing is applied, loss is computed in trainer's `compute_loss` function and `labels` are not shifted. This causes misalignment of `labels` and corresponding `input_ids`. This commit is for resolving this misalignment.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@sgugger

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2384,7 +2385,10 @@ def compute_loss(self, model, inputs, return_outputs=False):
             self._past = outputs[self.args.past_index]
 
         if labels is not None:
-            loss = self.label_smoother(outputs, labels)
+            if model._get_name() in MODEL_FOR_CAUSAL_LM_MAPPING_NAMES.values():
```
Comment (sgugger) [inline src/transformers/trainer.py:2388]: The `model` might need to be unwrapped here to deal with distributed settings.

**Review by sgugger — APPROVED** [review]
Thanks a lot for your PR! Left a small comment and it should be good to merge.
Make sure to run `make style` on your branch to apply formatting.

File: src/transformers/trainer.py (hunk 1 above)
Comment (seungeunrho) [reply src/transformers/trainer.py:2388]: Hi, thanks for the words. As you said, I've applied `make style`.
About the unwrapping `model` part, could you please elaborate? 
As for now, it seems there is no problem because the code passed all the tests.

Comment (sgugger) [reply src/transformers/trainer.py:2388]: Yes, the distributed tests are not run on each commit, so I would expect them to pass. You need to use the function `unwrap_model` (already imported in this module) or you might get `DistributedDataParallel` instead of your model name:
```suggestion
            if unwrap_model(model)._get_name() in MODEL_FOR_CAUSAL_LM_MAPPING_NAMES.values():
```

Comment (seungeunrho) [reply src/transformers/trainer.py]: Ah! Thanks for the explanation. I understood. 
I thought maybe just committing your suggestion is enough, so I did it.
